From 1bc024ea3b74664cd7bf3e493bd87011d274dc8b Mon Sep 17 00:00:00 2001
From: Ariel Zhang <arielzhang@google.com>
Date: Wed, 13 Mar 2024 17:19:21 +0000
Subject: [PATCH] Only extract components when necessary

Bug: b/327296393
Change-Id: Iabf48a1c413f33df3522809692f1cd33e2fde8b3
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5345512
Reviewed-by: Alexei Svitkine <asvitkine@chromium.org>
Commit-Queue: Ariel Zhang <arielzhang@google.com>
Reviewed-by: Luc Nguyen <lucnguyen@google.com>
Cr-Commit-Position: refs/heads/main@{#1272239}
---
 .../metrics/histograms/histogram_ownership.py |  3 ++-
 tools/metrics/histograms/merge_xml.py         | 23 +++++++++++++------
 .../histograms/print_expanded_histograms.py   |  2 +-
 tools/metrics/histograms/validate_format.py   |  2 +-
 4 files changed, 20 insertions(+), 10 deletions(-)

diff --git a/tools/metrics/histograms/histogram_ownership.py b/tools/metrics/histograms/histogram_ownership.py
index f3755a51ad1a1..d2d5e0c5ec553 100755
--- a/tools/metrics/histograms/histogram_ownership.py
+++ b/tools/metrics/histograms/histogram_ownership.py
@@ -70,7 +70,8 @@ def main():
   """
   if len(sys.argv) == 1:
     merged_xml_string = merge_xml.MergeFiles(
-        histogram_paths.ALL_XMLS, should_expand_owners=True).toxml()
+        histogram_paths.ALL_XMLS,
+        expand_owners_and_extract_components=True).toxml()
     root = ET.fromstring(merged_xml_string)
   else:
     rel_path = path_util.GetInputFile(
diff --git a/tools/metrics/histograms/merge_xml.py b/tools/metrics/histograms/merge_xml.py
index 2e643105ad23b..8156a10336541 100755
--- a/tools/metrics/histograms/merge_xml.py
+++ b/tools/metrics/histograms/merge_xml.py
@@ -206,28 +206,37 @@ def _BuildDOMTreeWithComponentMetadata(filename_or_file):
   return tree
 
 
-def MergeFiles(filenames=[], files=[], should_expand_owners=False):
+def MergeFiles(filenames=[],
+               files=[],
+               expand_owners_and_extract_components=False):
   """Merges a list of histograms.xml files.
 
   Args:
     filenames: A list of histograms.xml filenames.
     files: A list of histograms.xml file-like objects.
-    should_expand_owners: Whether we want to expand owners. By default, it's
-      false because most of the callers don't care about the owners for each
-      metadata.
+    expand_owners_and_extract_components: Whether we want to expand owners and
+      extract components. By default, it's false because most of the callers
+      don't care about the owners or components for each metadata.
 
   Returns:
     A merged DOM tree.
   """
   # minidom.parse() takes both files and filenames:
   all_files = files + filenames
-  trees = [_BuildDOMTreeWithComponentMetadata(f) for f in all_files]
-  return MergeTrees(trees, should_expand_owners)
+  trees = [
+      _BuildDOMTreeWithComponentMetadata(f)
+      if expand_owners_and_extract_components else xml.dom.minidom.parse(f)
+      for f in all_files
+  ]
+  return MergeTrees(trees,
+                    should_expand_owners=expand_owners_and_extract_components)
 
 
 def PrettyPrintMergedFiles(filenames=[], files=[]):
   return histogram_configuration_model.PrettifyTree(
-      MergeFiles(filenames=filenames, files=files, should_expand_owners=True))
+      MergeFiles(filenames=filenames,
+                 files=files,
+                 expand_owners_and_extract_components=True))
 
 
 def main():
diff --git a/tools/metrics/histograms/print_expanded_histograms.py b/tools/metrics/histograms/print_expanded_histograms.py
index 6b7766f7d628f..af15a6a94aa48 100755
--- a/tools/metrics/histograms/print_expanded_histograms.py
+++ b/tools/metrics/histograms/print_expanded_histograms.py
@@ -54,7 +54,7 @@ def main(args):
 
   # Extract all histograms into a dict.
   doc = merge_xml.MergeFiles(filenames=histogram_paths.ALL_XMLS,
-                             should_expand_owners=True)
+                             expand_owners_and_extract_components=True)
   histograms, had_errors = extract_histograms.ExtractHistogramsFromDom(doc)
   if had_errors:
     raise ValueError("Error parsing inputs.")
diff --git a/tools/metrics/histograms/validate_format.py b/tools/metrics/histograms/validate_format.py
index 1a53a031c4cc4..fa63624645c53 100755
--- a/tools/metrics/histograms/validate_format.py
+++ b/tools/metrics/histograms/validate_format.py
@@ -50,7 +50,7 @@ def CheckNamespaces():
 
 def main():
   doc = merge_xml.MergeFiles(histogram_paths.ALL_XMLS,
-                             should_expand_owners=False)
+                             expand_owners_and_extract_components=False)
   _, errors = extract_histograms.ExtractHistogramsFromDom(doc)
   errors = errors or CheckNamespaces()
   sys.exit(errors)
-- 
2.40.1

